docs/articles/documentation/guides/basic-guides/obtain-client-side-info.md

```diff
@@ -0,0 +1,172 @@
+
+> Do not modify the tested webpage within client functions.
+> Use [test actions](interact-with-the-page.md) to interact with the page instead.
+
+This topic contains the following sections.
+
+* [Client Function Constructor](#client-function-constructor)
+* [Run Asynchronous Client Code](#run-asynchronous-client-code)
+* [Execute Client Functions](#execute-client-functions)
+* [Overwrite Client Function Options](#overwrite-client-function-options)
+* [One-Time Client Code Execution](#one-time-client-code-execution)
+* [Import Functions to be Used as Client Function Dependencies](#import-functions-to-be-used-as-client-function-dependencies)
+* [Calling Client Functions from Node.js Callbacks](#calling-client-functions-from-nodejs-callbacks)
+* [Client Function Limitations](#client-function-limitations)
+
+## Client Function Constructor
+
+You can use the [ClientFunction](../../reference/test-api/global/clientfunction.md) constructor to create a client function.
+
+```js
+import { ClientFunction } from 'testcafe';
+const getWindowLocation = ClientFunction(() => window.location);
+```
+
+> Client functions cannot return DOM nodes. Use [selectors](select-page-elements.md) to achieve this.
+
+## Run Asynchronous Client Code
+
+TestCafe allows you to create client functions that run asynchronous code.
+
+You can pass a function that returns a Promise to the `ClientFunction` constructor to achieve this. In this instance, the client function will complete only when the Promise resolves.
+
+```js
+import { ClientFunction } from 'testcafe';
+
+const performAsyncOperation = ClientFunction(() => {
+    return new Promise(resolve => {
+        window.setTimeout(resolve, 500); // some async operations
+    });
+});
+```
+
+## Execute Client Functions
+
+Call the client function with the `await` keyword to execute it.
+
+```js
+import { ClientFunction } from 'testcafe';
+
+const getWindowLocation = ClientFunction(() => window.location);
+
+fixture `My fixture`
+    .page `http://www.example.com/`;
+
+test('My Test', async t => {
+    const location = await getWindowLocation();
+});
+```
+
+## Overwrite Client Function Option
+
+You can overwrite client function options with the ClientFunction's [`with`](../../reference/test-api/clientfunction/with.md) method.
+
+```js
+const cfWithDependency = cfWithoutDependency.with({
+    dependencies: { foo: 'bar' }
+});
+```
+
+## One-Time Client Code Execution
+
+You can use the [`eval`](../../reference/test-api/testcontroller/eval.md) method of the [test controller](../../reference/test-api/testcontroller/README.md) to create a client function and execute it immediately without saving it.
+
+```js
+fixture `My fixture`
+    .page `http://www.example.com/`;
+
+test('My Test', async t => {
+    const docURI = await t.eval(() => document.documentURI);
+});
+```
+
+## Import Functions to be Used as Client Function Dependencies
```

Comment:
Now the tricky part. There is a good way to deal with this section.

The thing is, this guide clearly lacks general information about client function dependencies (what we have in `options.dependencies` but guide-style).

I think we should create a subsection (I can't come up with a name, let it be *Call selectors and client functions from client-side code* for now). In there, we say about passing dependencies (Sometimes it's a good idea to reuse selectors and existing client functions' logic... - Andrey or Misha can also help you with real-life use cases if you want to add one), link `with` and constructor, and add an example.

Once we do this, this nasty `Import Functions to be Used as Client Function Dependencies` will go into the new section.